### PR TITLE
Synchronous Replication Infrastructure

### PR DESCRIPTION
**Pull request by tpan496**
# Synchronous Replication

## Description
Adds support for synchronous replication between noisepage clients. Supports TPCC.
## Remaining tasks

### CODE REVIEW

File: src/storage/recovery/recovery_manager.cpp [hunk 1]
```diff
@@ -58,11 +71,20 @@ void RecoveryManager::RecoverFromLogs() {
       }
 
       case (LogRecordType::COMMIT): {
+        //STORAGE_LOG_INFO("[Recovery Manager]: Processing COMMIT logs");
         NOISEPAGE_ASSERT(pair.second.empty(), "Commit records should not have any varlen pointers");
         auto *commit_record = log_record->GetUnderlyingRecordBodyAs<CommitRecord>();
 
         // We defer all transactions initially
-        deferred_txns_.insert(log_record->TxnBegin());
```
Comment (lmwnshn) [inline src/storage/recovery/recovery_manager.cpp:65]: Is there a reason for this?

File: src/storage/recovery/recovery_manager.cpp [hunk 2]
```diff
@@ -41,15 +43,26 @@ void RecoveryManager::WaitForRecoveryToFinish() {
...
   while (true) {
+    //STORAGE_LOG_INFO("[Recovery Manager]: Grabbing next log record...");
     auto pair = log_provider_->GetNextRecord();
     auto *log_record = pair.first;
 
     // If we have exhausted all the logs, break from the loop
-    if (log_record == nullptr) break;
+    if (log_record == nullptr) {
+      STORAGE_LOG_INFO("[Recovery Manager]: Exhausted all logs!");
+      if (exhausted_timer > 3) {
```
Comment (lmwnshn) [inline src/storage/recovery/recovery_manager.cpp:55]: What is the purpose of this timer?

Comment (tpan496) [reply src/storage/recovery/recovery_manager.cpp:55]: This is a hack to keep the recovery task keep running. I am not yet shutting it down correctly.

File: src/storage/recovery/recovery_manager.cpp (hunk 1 above)
Comment (tpan496) [reply src/storage/recovery/recovery_manager.cpp:65]: I didn't remove it, I moved it into the conditionals below.

**Comment by lmwnshn** [comment]
Different branch now.

**Pull request closed without merging**